### DISCUSSION TITLE
Implementation of high speed point-to-point communication (app-to-app)?

### DISCUSSION THREAD

**Discussion by pmehradnia** [nasa/cFS#425] [Q&A]
What are some different mechanisms to implement point-to-point communication to improve message passing performance? Mutex with OS_Queue is one option, but mqueues can be slow. What other solutions can be used to implement this?

**Comment by skliper** [comment]
Related: 
- https://github.com/nasa/CF/issues/193#issuecomment-1034191289
- https://github.com/nasa/osal/issues/73

@jphickey - did you have implementation options in mind?

**Reply by jphickey** [reply]
Yes - a ring buffer with a read and write index (integer).

So long as we use an atomic integer type for the index, and we restrict it (per API contract) that only a single reader and single writer thread are allowed to access the pipe at a time, this should avoid the need for internal mutex.

But such a design - while simple and fast - only really allows for polling.  Sometimes for FSW we want this in order to keep to the schedule (usually apps that are woken up via SCH shouldn't block on anything, just wait till next cycle).

To support immediately/actively waking up of a reader thread when new data is available, we'd need to have a sync mechanism.  This is where I'd use pthread condition variables (ideal for this type of thing) but OSAL does not expose these at this time.

**Reply by pmehradnia** [reply]
Interesting. Basically, reader would only read if write index is past the read index, and vice versa for write. And the buffer can simply be a contiguous C array. Would you place this as part of a library with a static or global internal buffer, and have the two different cFS Apps call the library API function (say, parameterized to indicate _Write_ or _Read_) defined in the library directly?

**Reply by skliper** [reply]
Grab from the ES memory pool?  Allows for creation/deletion as needed within the bounds of the pool (same is used for SB messages).

**Reply by pmehradnia** [reply]
Yes that would be a good option so it is managed by ES. But you would still keep it within a library context, accessible by the two cFS Apps?

**Reply by skliper** [reply]
Library or provided by the framework, yes (maybe OSAL since it has the queue APIs?  Seems like this is basically what's requested in nasa/osal#73 and could just extend what's already there if needed?).  Likely associated with an ID.  The writer could own/create, and reader could GetIdByName.

**Reply by jphickey** [reply]
SB already has its own memory pool and buffers for this can be allocated from the same pool and even follow the same general pattern as a "zero copy" message (CFE_SB_AllocateBuffer, etc). It can be pulled from a different memory pool if need be, but I think there are enough similarities that I'd rather try to keep it the same rather than making it different upfront, and only make it different if some necessity arose.

I think the basic Ring Buffer concept should live in OSAL ... the code itself would be mostly portable/shared but it may eventually want some sync hooks that would need to call into the Impl layer. The Ring Buffer API would look a lot like the current Queue API, so for SB I see it as an option when creating a pipe (e.g. a POINTTOPOINT flag to CFE_SB_CreatePipe?) that causes it to create this ring buffer instead of a conventional msgqueue. The resource semantics would all apply - it gets an OSAL ID and all typical OSAL patterns apply.

The main difference would be on the SB side of things - actually it is a simplification. If SB knows that this pipe is supposed to only be one publisher and one subscriber, it can bypass a lot of the logic to do "broadcasting" that is currently in there. It can also propagate back-pressure if the reader isn't draining the ring as fast as the writer is putting it in.

Challenge is to figure out how much we should trust the developer and how much we should verify that they did what they said they said they would do. For example it could verify that it is indeed only being accessed by one thread at a time but that adds complexity and reduces the speed by requiring sync primitives and all. If we want it to be fast and efficient, then the answer may be to just document that it only works with single reader and single writer, and if the developer violates this rule, results are undefined.

**Reply by skliper** [reply]
I think it's worth trading the potential confusion that may be introduced by adding it to Software Bus, since that module has become synonymous with pub-sub.  If it's just a point to point queue I feel like it could be a very simple library/module with a single purpose and other than an API to report it's statistics really wouldn't require the complexities of the SB cmd/tlm interface or management task (or potentially require modifications to that interface).  Seems easier than cracking open SB and avoids potential pitfalls of coupling with an existing, already complex module.  I think we should strive for simplicity and efficiency, with appropriate documentation.

**Reply by jphickey** [reply]
Yeah, understood ... but the flip side of that is the fact that SB has been the historical provider of the "messaging interface" in the more general sense.  That is, when documenting the "inputs" and "outputs" of apps, it is currently in terms of MsgId (or "Topic IDs" is a term I like better, but the same idea); that is, this topic ID has this type of information associated with it.

In particular, I think there is value in keeping the similar paradigm, where all inputs and outputs of an app are via MIDs/TopicIDs, not some other new/separate message passing facility that users will not be accustomed to.  Since apps like CF are today sending and receiving data PDUs via MIDs, keeping these P2P interfaces as similar to MIDs as possible makes sense to me, it would minimize breakage to those apps.  

In particular I think a lot of the same status reporting that SB does on its current pipes, probably also applies to this P2P concept.  They'd still have current depths, high watermarks, soft and hard limit events, etc.  There are a lot of similarities/overlap to current SB pipes where making a separate service would necessitate duplicating all of that.

**Reply by skliper** [reply]
I'd think P2P could be completely abstracted from what's actually transferred other than an address and size provided by the caller (or just a handle w/ that info tracked if memory was provided by something like allocate buffer).  I don't see the need for the API's to care about MIDs, the apps could own all of that logic (what MIDs go in the queue, format of the data in general, etc).  CF would still put the PDU's in CCSDS packets (zero change there) and the receiver is still getting CCSDS packets out (zero change there) it's just the transfer mechanism that changes to take advantage of the fact no special routing/synchronization/etc is necessary.  Maybe what I have in mind is smaller scope than what you are thinking?  This doesn't actually solve the issue where variable sized CCSDS packets may not be handled well by the receiving peer (or downstream), but that's a different problem than implementing an efficient P2P interface (whatever is downstream just needs to be updated to handle variable size or the source needs to produce fixed size packets.)

**Reply by skliper** [reply]
In other words, I see MIDs as the efficient way of identifying CCSDS packets for routing, filtering, etc.  I'm suggesting all that logic be separate from the actual P2P interface... the less it cares about the more efficient it can be.

**Reply by jphickey** [reply]
Actually, in a P2P context, I do not think CF would put data into CCSDS packets at all.  The sender (CF) would know its a point to point link and it can skip all that extra work.  The only reason I can see it still using CCSDS framing would be as a compatibility stop-gap measure in case one or the other side of a connection did not know it has a P2P link (but if that's the case then backpressure won't be possible either).

I was referring to MIDs here not for how the software uses it, but how a systems engineer uses it.  In the concept of "this data exists in my system, how do I get it to another part of my system that is interested in it".  In the case of CFDP PDUs, it's a matter of connecting CF to something that can send those PDUs to their next hop on the eventual path to their destination.  The system in general just needs a way to connect the producer to the consumer, in an abstract way, that doesn't reinvent the wheel.  SB currently does this via MIDs.a  Perhaps it could (and maybe should?) use text-based names instead for ALL of this, it would be much more abstract.  (i.e. even in a pub/sub context, instead of subscribing to `SAMPLE_APP_HK_TLM_MID` ... it could subscribe to `cfe_sb://SAMPLE_APP:1/HK_TLM` -- made to look URI-like).

In a P2P context, there is no routing on that hop, so CCSDS framing doesn't add much value.  Furthermore, CCSDS has other problems in this context - a big one is that it insists that the data be labeled as "CMD" or "TLM" -- its a fixed binary bit that has to be set one way or the other - which may lead to misclassification/mishandling of the data - exactly the problem we need to avoid, as this is neither CMD nor TLM.

IMO the root of this problem is the fact that the system labels these CFDP PDUs as "Telemetry data" ... which by definition is historically fixed size (not even a terrible assumption to make, TLM _is_ supposed to be fixed/consistently formatted messages after all).  So for one hop to want to assign a "known size" to a given TLM MID is not wrong.  What's wrong is the fact that it was labeled as TLM data in the first place (its not).

**Reply by jphickey** [reply]
To summarize - I would definitely be in favor of replacing numeric MIDs entirely (both for existing pub/sub as well as for future p2p) with a more abstract endpoint name like a CFE URI or similar.  Those endpoint names can be looked up in a the mission data dictionary (because we have one, right?) that reveals the nature of the data (its size, frequency, nature of the data, its routing info i.e. the MID it is actually being published on).  Such a lookup would only be done at time of subscription which is really just a start up thing, not for every message, of course.  

In the meantime, without that separate lookup table, we've historically used numeric MID for that purpose, making that number do double-duty as both an administrative identifier and runtime identifier.  The latter (runtime) is not necessary for P2P, but the former (admin) is.

**Reply by skliper** [reply]
Although I agree CF wouldn't have to put PDUs into CCSDS packets, something does.  I see more negatives in pushing that work downstream than positives (one less header would be slightly less overhead in the queue but the extra logic downstream to handle the CF P2P link as "special" doesn't seem worth it).  As long as it's CCSDS packets, everything downstream knows how to handle it using already established techniques, and as long as the downstream entities actually look at the size it's just like any other CCSDS packet when it comes to routing and there doesn't need to be any unique logic.

I still don't think the P2P interface software or system engineers need to care what MIDs are on the link.  It's simply a queue the way I think of it.  There's nothing that requires CCSDS packets to be fixed size other than historical systems that can't handle variable sized CCSDS packets.  Although I do see the possible confusion in labeling PDU's as TLM since they can be received or sent I really don't see that as an issue... it's not an app command by any means, it's on a unique MID so I'm fine with calling it TLM.  Other apps receive TLM and handle it appropriately (think DS, HS, HK, etc all receive tlm), I don't see why CF can't do the same.  To every other entity but CF it's just a CCSDS packet with an MID that needs to be routed to the appropriate destination.  For incoming the external link could simply filter on MID's (not caring if it's marked as TLM or CMD) and put it on the P2P to CF.  Outgoing it gets the CCSDS packet from CF over the P2P and simply puts it on the external link.  Much like every other CCSDS packet, only the endpoint cares if it's labeled CMD or TLM.  Although FWIW there are implementations that may handle front end CMD verification... but as long as it's labeled TLM there's no impact.

Summary - Value of CCSDS prior to the P2P hop keeps the unique handling localized.  If the other endpoint is just handling CCSDS packets impacts are minimized, then P2P becomes useful in many scenarios where efficiency is a priority over pub/sub.  As soon as you start transmitting raw PDU's across an interface, anything that receives it is now required to know what a PDU is.

**Reply by skliper** [reply]
All good conversation for redesign options of CF and PDU handling, MIDs and CCSDS handling redesign, etc but maybe a good time to move those topics back to their more closely associated issues (maybe https://github.com/nasa/CF/issues/193 and https://github.com/nasa/cFE/issues/732?)  Back to the P2P interface, my preference is to not bake it into SB and provide as either an extension of OS_Queue's or a separate library with a documented point to point restriction (design for efficiency and support the concept of back-pressure) that treats the data as opaque (size and address).  Then it's generally applicable to a wide variety of use cases whatever path we go down for CF.  Really could just be a user-space queue limited to P2P which could be a subset of or related to https://github.com/nasa/osal/issues/73?
